test/onnx/test_utility_funs.py

```diff
@@ -25,6 +27,115 @@ def forward(self, x):
+                return b + x
+
+        _set_opset_version(9)
+        x = torch.ones(3, 2)
+        graph, _, __ = utils._model_to_graph(TransposeModule(), (x, ), None,
+                                             do_constant_folding=True,
+                                             _disable_torch_constant_prop=True)
+        for node in graph.nodes():
+            assert node.kind() != "onnx::Transpose"
+            assert node.kind() != "onnx::Cast"
+            assert node.kind() != "onnx::Constant"
+        assert len(list(graph.nodes())) == 1
+
+    def test_constant_fold_slice(self):
+        class SliceModule(torch.nn.Module):
+            def forward(self, x):
+                a = torch.tensor([[1., 2., 3.], [4., 5., 6.]])
+                b = torch.narrow(a, 0, 0, 1)
+                return b + x
+
+        _set_opset_version(9)
+        x = torch.ones(1, 3)
+        graph, _, __ = utils._model_to_graph(SliceModule(), (x, ), None,
+                                             do_constant_folding=True,
+                                             _disable_torch_constant_prop=True)
+        for node in graph.nodes():
+            assert node.kind() != "onnx::Slice"
+            assert node.kind() != "onnx::Cast"
+            assert node.kind() != "onnx::Constant"
+        assert len(list(graph.nodes())) == 1
+
+    def test_constant_fold_unsqueeze(self):
+        class UnsqueezeModule(torch.nn.Module):
+            def forward(self, x):
+                a = torch.tensor([[1., 2., 3.], [4., 5., 6.]])
+                b = torch.unsqueeze(a, 0)
+                return b + x
+
+        _set_opset_version(9)
+        x = torch.ones(1, 2, 3)
+        graph, _, __ = utils._model_to_graph(UnsqueezeModule(), (x, ), None,
+                                             do_constant_folding=True,
+                                             _disable_torch_constant_prop=True)
+        for node in graph.nodes():
+            assert node.kind() != "onnx::Unsqueeeze"
+            assert node.kind() != "onnx::Cast"
+            assert node.kind() != "onnx::Constant"
+        assert len(list(graph.nodes())) == 1
+
+    def test_constant_fold_concat(self):
+        class ConcatModule(torch.nn.Module):
+            def forward(self, x):
+                a = torch.tensor([[1., 2., 3.]])
+                b = torch.tensor([[4., 5., 6.]])
+                c = torch.cat((a, b), 0)
+                return b + c
+
+        _set_opset_version(9)
+        x = torch.ones(2, 3)
+        graph, _, __ = utils._model_to_graph(ConcatModule(), (x, ), None,
+                                             do_constant_folding=True,
+                                             _disable_torch_constant_prop=True)
+        for node in graph.nodes():
+            assert node.kind() != "onnx::Concat"
+            assert node.kind() != "onnx::Cast"
+            assert node.kind() != "onnx::Constant"
+        assert len(list(graph.nodes())) == 1
```

Comment:
@ezyang - I would love to be able to improve this test point (or constant folding for that matter), but I am not sure I fully understand the question. 

First, as I asked in https://github.com/pytorch/pytorch/issues/28834 also (I saw that issue before I saw your comment here), I am not sure I understand what is meant by initialization graph. 

The goal of this test is simple. I have a graph where I have two constants feeding into a `Concat` node, followed by an `Add` op. Regarding the `Cast` op, I recall that they were not intentionally part of this test, but appeared in the graph when we set `_disable_torch_constant_prop=True`. However, since `Cast` is supported in constant folding they get folded too. Anyway, this test was supposed to test that the `Concat` node is constant folded (together with `Cast` op). The new folded constant after collapsing the `onnx::Constant` and the `onnx::Concat` node shows up as onnx initializer and not `onnx::Constant` in the final ONNX graph. This is why we check that there are no `onnx::Constant`, cast, or concat nodes. 

As an aside - looking back at this I feel this test would be better if I use the input `x` to the `forward()`  method in the model. I think it will make things clearer. I will try to update the test.

Reply:
> what I expect to happen here is to get a graph with a single node that returns a constant. 

Forgot to mention - we do not support `Add` in constant folding yet, but if we did you are right in that the Add op should have been folded too. You would be left with a single intializer in the final ONNX graph.